expyfun/_experiment_controller.py

```diff
@@ -1101,6 +1106,57 @@ def check_force_quit(self):
         """Check to see if any force quit keys were pressed."""
         self._response_handler.check_force_quit()
 
+    def text_input(self, stop_key='return', pos=[0, 0], color='white',
+                   font_name='Arial', font_size=24, wrap=True, units='norm',
+                   all_caps=True):
+        """Allows participant to input text and view on the screen.
+
+        Parameters
+        ----------
+        stop_key : str
+            The key to exit text input mode.
+        pos : list | tuple
+            x, y position of the text. In the default units (-1 to 1, with
+            positive going up and right) the default is dead center (0, 0).
+        color : matplotlib color
+            The text color.
+        font_name : str
+            The name of the font to use.
+        font_size : float
+            The font size (in points) to use.
+        wrap : bool
+            Whether or not the text will wrap to fit in screen, appropriate
+            for multi-line text. Inappropriate for text requiring
+            precise positioning or centering.
+        units : str
+            Units for `pos`. See `check_units` for options. Applies to
+            `pos` but not `font_size`.
+        all_caps : bool
+            Whether the text should be displayed in all caps.
+        """
+        letters = string.ascii_letters
+        stop = False
+        text = str('')
+        while not stop:
+            letter = self.wait_one_press(timestamp=False)
+            if letter in letters and all_caps:
+                letter = str.upper(letter)
+            if letter == stop_key:
+                stop = True
+            else:
+                if letter == 'backspace':
+                    text = text[:-1]
+                elif letter == 'space':
+                    text = text.ljust(len(text) + 1)
+                elif letter in letters:
+                    text += letter
+                self.screen_text(text + '|', pos=pos, color=color,
+                                 font_name=font_name, font_size=font_size,
+                                 wrap=wrap, units=units, log_data=False)
```

Comment:
Can this be simplified to:
```
while True:
    letter = self.wait_one_press(timestamp=False)
    if letter == stop_key:
        break
    if letter == 'backspace':
        text = text[:-1]
    else:
        letter = ' ' if letter == 'space' else letter.upper()
        letter = letter.upper() if all_caps else letter
        text += letter
    self.screen_text(...)
    self.flip()
```
?

Reply:
(One change that might matter is that here I have removed the check for `letter in letters` because e.g. `'-'.upper() == '-'`)

Reply:
ooh, I like that better.

Reply:
the space is an issue because it just reads as an empty string

Reply:
wait.  the `break` means that `self.flip()` won't get called when `stop_key` is pressed, and the text will just stay there on screen.

Reply:
Nevermind, apparently that's not an issue anymore?

Reply:
easy enough to solve though:
```
...
if letter == stop_key:
    self.flip()
    break
...
```

Reply:
sure, add it there, add an extra branch level, or add it to the end after the `while ` loop is over

Reply:
Ahh so Pyglet doesn't give characters, it gives human-readable names. Bummer